bus-mapping/src/circuit_input_builder.rs

```diff
@@ -1118,6 +1121,117 @@ impl<P: JsonRpcClient> BuilderClient<P> {
         Ok((proofs, codes))
     }
 
+    /// Yet-another Step 3. Get the account state and codes from pre-state tracing
+    /// the account state is limited since proof is not included,
+    /// but it is enough to build the sdb/cdb
+    /// if a hash for tx is provided, would return the prestate for this tx
+    pub async fn get_pre_state(
+        &self,
+        eth_block: &EthBlock,
+        tx_hash: Option<H256>,
+    ) -> Result<
+        (
+            Vec<eth_types::EIP1186ProofResponse>,
+            HashMap<Address, Vec<u8>>,
+        ),
+        Error,
+    > {
+        let traces = if let Some(tx_hash) = tx_hash {
+            vec![self.cli.trace_tx_prestate_by_hash(tx_hash).await?]
+        } else {
+            self.cli
+                .trace_block_prestate_by_hash(
+                    eth_block
+                        .hash
+                        .ok_or(Error::EthTypeError(eth_types::Error::IncompleteBlock))?,
+                )
+                .await?
+        };
+
+        let mut account_set =
+            HashMap::<Address, (eth_types::EIP1186ProofResponse, HashMap<Word, Word>)>::new();
+        let mut code_set = HashMap::new();
+
+        for trace in traces.into_iter() {
+            for (addr, prestate) in trace.into_iter() {
+                let (_, storages) = account_set.entry(addr).or_insert_with(|| {
+                    let code_size =
+                        Word::from(prestate.code.as_ref().map(|bt| bt.len()).unwrap_or(0));
+                    let (code_hash, keccak_code_hash) = if let Some(bt) = prestate.code {
+                        let h = CodeDB::hash(&bt);
+                        // only require for L2
+                        let keccak_h = if cfg!(feature = "scroll") {
+                            hash_code_keccak(&bt)
+                        } else {
+                            h
+                        };
+                        code_set.insert(addr, Vec::from(bt.as_ref()));
+                        (h, keccak_h)
+                    } else {
+                        (CodeDB::empty_code_hash(), *KECCAK_CODE_HASH_EMPTY)
+                    };
```

Comment:
i hope the code hash calculation will not be run during ccc?   poseidon hash in very slow.  In current ccc, there is (almost?) no code hash.  All code hash will be provided inside trace.

Reply:
oh i remember this "get_pre_state" is not related to scroll evm?  We will use storage_proof inside trace?